### ISSUE TITLE
Reuse input lock files

### ISSUE THREAD

**Issue by roberth** [NixOS/nix#7730]
**Is your feature request related to a problem? Please describe.**

Lock files contain more nodes than needed.
This leads to an incentive to put inputs that are not for consumers of a flake elsewhere, degrading the developer experience.
Typically this involves a second flake lock in a subdirectory for development dependencies, which needs to be managed separately, may go out of sync with the root flake, etc.

**Describe the solution you'd like**

Only store the nodes that are direct inputs or those where it makes sense to store a copy, when a `follows` makes it possible to access the *otherwise* transitive input as a direct input.
This way, we have no loss of performance.
"Missing" transitive nodes are sourced from the inputs' lock files that contain them.

**Status**

Not implemented yet.
[Migration plan below](https://github.com/NixOS/nix/issues/7730#issuecomment-3568359904).

<details><summary>Previous issue description, initial substantiation</summary>

(was: Solve dev(/test) dependencies and bloated, diverging lock files, by reconsidering the flattened lock file)

**Is your feature request related to a problem? Please describe.**

We've previously identified issues with the way the lock file currently works.

 - Irrelevant development dependencies are added to the lock files of all dependent flakes
    - https://github.com/NixOS/nix/issues/6124
 - Mututally dependent flakes cause diverging lock files as each lock file update incorporates a new version with a chain of irrelevant dependencies.
    - _issue?_
 - Some users just complain about the size of their lock file. Partly pedantic, but they're actually kind of right:
 - Ideally, lock file changes are reviewable, so that nothing weird creeps in through contributions that have such updates.

**Describe the solution you'd like**

We may consider two scenarios; one with `follows` and one without. Let's start without `follows`.

Observing that we do not exploit the apparent benefit of knowing all inputs upfront, we may consider loading the required parts of dependency lockfiles on demand.

Example scenario
 - Dependencies `.` -> `foo` -> `nixpkgs`
 - No `follows`. 

Evaluation order, current situation:
 - need `packages.default`
 - need `inputs.foo.packages.foo`
 - read `foo` details from `./flake.lock`
 - fetch `foo`
 - need `foo.inputs.nixpkgs`
 - read `nixpkgs_1` details from local lock file
 - fetch `nixpkgs_1`

Same evaluation, partial lock file
 - need `packages.default`
 - need `inputs.foo.packages.foo`
 - read `foo` details from local lock file
 - fetch `foo`
 - need `foo.inputs.nixpkgs`
 - read `nixpkgs` details from `foo/flake.lock`
 - fetch `nixpkgs`

As you can see, the fetching characteristics remain the same.

Let's reintroduce `follows`, and first consider the example:

```
inputs.foo.nixpkgs.follows = "nixpkgs";
```

In other words, this injects the local `nixpkgs` dependency into `foo`. The lazy fetching behavior remains the same. This injection should be handled by `call-flake.nix` and does not require changes to the lock file.

Now consider the opposite `follows`:

```
inputs.nixpkgs.follows = "foo.nixpkgs";
```

In this case, if we were not to include a copy of `foo.nixpkgs` details in the local lock file, we'd have a slightly worse lazy fetching behavior, as `nixpkgs` would only be fetchable after `foo`'s lock has been fetched. For this case, we do want "flat lock file" behavior, but limited to this dependency. The rule for which transitive dependencies to include in the lock file might be as simple as _direct inputs + inputs referenced in `follows`_.

This suggests the following implementation strategy:
 - When generating a lock, apply the above rule to determine which flakes to traverse and include in the lock file
 - When calling a flake, call it such that its `inputs` are _those from its own lock_ `//` _those set in the current lock_
    - https://github.com/NixOS/nix/pull/6530 already adds an `overrides` parameter in `call-flake.nix` that appears to enable this. The missing bit is that some/many nodes would be created by reinvoking `call-flake.nix` rather than sharing the node identifiers. Node identifiers are local to each `call-flake.nix` call. i.e. `overrides` is a prerequisite for calling dependency locks.

**Describe alternatives you've considered**

Mark development dependencies explicitly.
 - This is _more configuration_ which is generally not good.
 - This adds more corner cases and error messages.
 - This does not solve the problem of having mutually recursive development dependencies. Note that we can barely manage to make builds not mutually recursive, but flakes encompass more than builds, so expecting flakes not to be mutually recursive is not realistic.

</details>

**Additional context**

- Alternative or complement to https://github.com/NixOS/nix/issues/6124

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by roberth** [comment]
I quite like the syntax @mkenigs used in #4004 to talk about overrides

```
A (override B.C.D.E) -> B (override C.D)-> C -> D -> E
```

It is also a useful vehicle for talking about the locking algorithm, which seems that it would recurse into dependencies until the overrides are `()`, and would only record the inputs that are not already locked or overridden somewhere.

Similarly, it would probably reflect the `overrides` parameter in `call-flake.nix`. This parameter does not exist in `master` yet, and `lazy-trees` defines it as overrides on sources. It would have to become a structure representing overrides containing invoked flakes rather than source inputs, and it perhaps need a somewhat richer structure to represent deep overrides.

**Comment by edolstra** [comment]
@tomberek My thought on the new lock file format: It should store mappings from unlocked input paths to locked refs, and nothing more, and it shouldn't store input paths that are already locked by the lock files of input flakes.

I.e. something like

```
{
  "version": 8,
  "locks": {
    "nixpkgs": {
      "original": { ... original ref from flake.nix ... }
      "locked": { ... the immutable lock ... }
    },
    "nix": {
      "original": { ... original ref from flake.nix ... }
      "locked": { ... the immutable lock ... }
   },
   # note: since the "nix" input has its own lock file, we don't store the locks of the "nix" input here, unless there are overrides
   # e.g. if there is a (non-follows) override "nix.nixpkgs", or if "nix" does not have a lock file:
   "nix.nixpkgs": {
     "original": ...,
     "locked": ...
   };
  }
} 
```

Thus, if we encounter an unlocked input `a.b.c` in flake.nix, we look up `locks."a.b.c"` in `flake.lock`. If it doesn't exist (or if its "original" ref doesn't match), we resolve the unlocked ref and add/replace the lock file entry `a.b.c`.

We don't create lock file entries for "follows" inputs (e.g. if "a.b" follows "c.d", we don't create a lock file entry for "a.b"). In general, we don't store any information in the lock file that is already in flake.nix (except for the "original" refs, which are needed to decide whether we need to update a lock file entry if the corresponding unlocked entry has changed).

Note that the above replaces the graph structure of the current lock file format with a simple tree.

**Comment by roberth** [comment]
> Note that the above replaces the graph structure of the current lock file format with a simple tree.

- current graph: everything, ready to for a `mapAttrs` and `fix` without doing a recursive `call-flake.nix` - only one set of nodes
- proposed tree: just the impure flakerefs in the current flake's inputs. `call-flake.nix` does something similar to the current `call-flake.nix`, but does so for individual flakes with their inputs, without the unnecessary concept of nodes.

To make `follows` possible, `call-flake.nix` needs to accept input overrides as a parameter. Without such a parameter, an input A would only see its own inputs, because `call-flake.nix` reinvoked itself with A's lock file instead of the user's lock file.

**Comment by tomberek** [comment]
From NixCon, in discussion with @lheckemann , @infinisil , @roberth 

- established an example where behavior differs. When an immediate dependency moves or is no longer available (eg: authorization), flakes it depends on are not accessible. This means something like `inputs.A.inputs.B.inputs.C.packages.SYSTEM.thing` would fail, where with a full-graph would allow for accessing C's outputs even if B was unavailable. So in this particular sense we lose this style of laziness. This is an edge-case and likely not worth basing decision-making on.
- Simpler implementation is highly preferable.
- Need a strategy to still support older lockfiles? (for how long? until stabilization?)
- Need for a cache of resolved lockfile information
    - can we use existing git cache?
    - reuse eval cache sqlite?
    - another new cache?
- include a strict-lock option?

**Comment by mkenigs** [comment]
One potential consideration for the new format:

After @lheckemann's talk I was thinking about a locking algorithm that would handle something like this:
flake A's flake.nix:
```
inputs.B.url = "github...";
inputs.B.useLock = false;

inputs.C.url = "github...";
inputs.C.useLock = false;
```

Suppose B and C both depend on nixpkgs. Then locking flake A would store a single revision for nixpkgs and store it in flake A's flake.lock. This is very similar to how language dependency managers like cargo make a distinction between lib and binary dependencies.

It might be possible to express the same thing in the current format, by doing e.g.:
```
inputs.nixpkgs = "github...";
inputs.B.inputs.nixpkgs.follows = "nixpkgs";
inputs.C.inputs.nixpkgs.follows = "nixpkgs";
```

But I think that's more difficult, and in practice it seems people often don't add all of the follows they could.

Implementing `useLock` would require storing the entirety of the dependency graph of the child in the parent's flake.lock. That's possible in the current format. It's not possible with the format currently proposed in this issue, although I think it could be a natural extension (i.e. `useLock` could imply the strict-lock option @tomberek mentioned)

**Comment by lheckemann** [comment]
It should in fact be possible with the proposed format, and would be represented on the lockfile side in much the same way as the `follows` approach.

Note also that ignoring downstream locks would only result in the same nixpkgs being used if the downstream inputs use _the same_ nixpkgs input spec. I hope we can also find an approach (though it need not be implemented inside Nix) that works for replacing references that use different branches, or even different repositories or input types.

**Comment by roberth** [comment]
> would be represented on the lockfile side in much the same way as the `follows` approach.

Yes, except `follows` doesn't have to be in the lockfile at all. We expect that the `inputs` declarations in `flake.nix` can be the single source of truth. Same for features that behave like `follows`.

**Comment by roberth** [comment]
> (it's a bit more like an implicit follows).

An implicit follows is an override just as well.

-----


Also responding to your previous comment:

> Implementing `useLock` would require storing the entirety of the dependency graph of the child in the parent's flake.lock. That's possible in the current format. It's not possible with the format currently proposed in this issue

It is actually possible, by storing the whole graph as overrides. `useLock` would cause `call-flake.nix` to look for, and apply those overrides.

So such a feature does not conflict with the proposal to reuse input lock files without copying.

> although I think it could be a natural extension (i.e. `useLock` could imply the strict-lock option @tomberek mentioned)

Yes, except applied to part of the dependency tree. It's even more similar to what we do for inputs that have no lock file.

For `useLock` to work well, it would be nice to have inputs that are "fixed" or "dev only" or "private" - ones that aren't affected by `useLock` because that wouldn't achieve anything.

This can all be added later though.

**Comment by mkenigs** [comment]
> It is actually possible, by storing the whole graph as overrides. `useLock` would cause `call-flake.nix` to look for, and apply those overrides.

I think I may be misunderstanding:
> proposed tree: just the impure flakerefs in the current flake's inputs

Which I took to mean inputs/overrides/follows mentioned in flake.nix. But if storing the whole graph as overrides, including inputs that aren't in any way present in flake.nix, is possible, then that's great.

> This can all be added later though.

👍 for sure! If I do take a crack at implementing `useLock`, should I do that using the existing format, or do the changes mentioned in this issue need to happen first?

**Comment by szlend** [comment]
Another issue with the current lock file is that it causes nix to needlessly re-evaluates equivalent inputs.

For example if we have two inputs `A` and `B` that both depend on the same revision of `nixpkgs` (equivalent `narHash`), then `nixpkgs` will be evaluated twice. This can be worked around by using `B.inputs.nixpkgs.follows = "A/nixpkgs"`.

However if you want to take advantage of binary caching, then using `follows` is not an option. If `A.nixpkgs` and `B.nixpkgs` at some point diverge because they aren't always updated at the same time, then using `follows` on `B.nixpkgs` will cause it to miss the binary cache.

So if you want to take advantage of binary caching, there's no practical way to avoid re-evaluating inputs.

**Comment by szlend** [comment]
I looked into fixing the performance implications of this (without touching the lock file format) in [#10218](https://github.com/NixOS/nix/pull/10218).

**Comment by roberth** [comment]
@szlend Thank you for looking into this.

I think the solution in #10218 and in this proposal are somewhat incompatible, because #10218 would require fetching all lock files all the time. It seems that laziness could be recovered with #8025 though, so I'm not too worried.

**Comment by roberth** [comment]
- https://github.com/edolstra/flake-compat/pull/49 appears to implement at least some of the overriding logic (though it's a useless diff, even if the code may be great, idk)

**Comment by onnimonni** [comment]
I came here to follow this feature from the discussions of https://github.com/NixOS/nix/pull/10089#issuecomment-2575158032:

> > An update to the input of a subflake does not get automatically picked up by the root when running a build.

> That's right, and probably not ideal, but it's consistent with how lock files currently work in general: lock files are independent, and the only time that we use the lock file of an input is if the referring flake doesn't have a lock for that input yet. So a change to the lock file of an input doesn't propagate to the referring flake.
> This should indeed be fixed by having sparse lock files (https://github.com/NixOS/nix/issues/7730).

This is rather inconvenient but I can for now use the `override-input` flag:
```
--override-input subflake ./subflake/
```

It would be amazing if multiple flakes inside a monorepo would be in sync without using that or running `nix flake update subflake`.

Current pretty weird behaviour was documented well in https://github.com/NixOS/nix/pull/10089#issuecomment-2561547950.

**Comment by roberth** [comment]
I've done a bit of exploration and planning to hopefully help this forward a bit.
I haven't gone in depth about the update algorithm yet, but I have explored how the migration could be managed.

# 10,000 foot overview

- Add `sparseNodes` field to root of lock file
  - referencing the existing nodes for their source pins
  - purpose: augment the `flake.nix` `inputs` structure (expecting close to 1:1 correspondence until we learn more)
- Provide opt-in to make this `sparseNodes` and `flake.nix` the source of truth instead of the nodes. Nodes only provide source info.
- Introduce feature and validate the feature with care

# Requirements

- Goal: Reduce lock file bloat by storing only direct inputs in sparseNodes, falling back to child flakes' lock files for transitive dependencies.
- Minimize impact of solution validation on ecosystem
- De-risk the change as soon as possible
  - Implement whole migration before releasing phase 1
  - Test end-state behavior asap, balanced with other requirements
- Avoid risk of stale lock data
- Allow gradual validation and adoption
  - Keep flake lock files backward compatible until the very end of the migration - when feature is validated and adoption is very high

# Findings

  - Old Nix removes unknown `sparseNodes` root attribute when regenerating locks (correct - prevents stale data)
  - Preserved only when lock already matches input declarations
  - Gradual adoption: more sparseNodes as ecosystem upgrades
  - Some "flickering" when lock is maintained using multiple Nix versions. Not great, but not too intrusive and can be improved by adding explanatory comment in lock file.

# Design Decisions

  1. Always Generate `sparseNodes` - Unconditionally, regardless of feature flag. Essential for ecosystem-wide testing since users can't regenerate all their dependencies' locks.
  2. Reference-Based Format (Phase 1) - sparseNodes stores references to nodes entries, not duplicates. Eliminates data divergence risk.
  3. Parse-Time Resolution - References resolved immediately by copying data, creating self-contained sparse tree. Tests end-state behavior from day one.
  4. Develop All Phases Before Release - Implement phases 1-4 before releasing Phase 1. Enables incremental adoption/validation without incremental development risk.
  5. Self-Documenting - Include /comment field inside sparseNodes linking to documentation.

# Release plan

  | Phase | Description                                                             | Breaking?        |
  |-------|-------------------------------------------------------------------------|------------------|
  | 1     | Generate both `nodes` + `sparseNodes` (references), consumption behind flag | No               |
  | 2     | Community testing, opt-in `sparseNodes` consumption                                   | No               |
  | 3     | Default enabled, opt-out `sparseNodes` consumption                                    | No               |
  | 4     | Remove `nodes` from generated locks, sparse-only, sources by value                                        | Yes (far future) |

Breaking refers to the ability of older Nix versions to evaluate such flakes. Lock `version` should be incremented at that point.

# Risk mitigations

  - Data divergence: reference-based storage until phase 4
    - Switch to storing `sparseNodes` sources _by value_ in `sparseNodes` itself is trivial but must be tested early
  - Stale data: old Nix removes unknown fields
  - Development risk: implement all phases built before release of phase 1

# Remaining Work

  - Design the `follows` detection algorithm
  - Specify sparse resolution semantics
  - Implementation of all phases
  - Test suite

**Comment by roberth** [comment]
Maybe it should just be `inputs` instead of `sparseNodes` if they can be in a 1:1 correspondence with the `flake.nix` `inputs`.

**Comment by roberth** [comment]
~~This issue is currently part of the flakes stabilisation milestone, which I believe is unnecessary.~~
See next comment and disregard this one.

It would be _comfortable_ to solve this during the experimental phase, but _out of necessity_ you need things like compatibility measures and a careful rollout anyway, so you don't get much actual benefit from the experimental phase for this one.

My understanding of the stabilisation milestone is that it focuses on changes that either
1. are hard to gauge because their implementation (or "specification") isn't well understood
2. are known to cause a breaking change, such as evaluation changes (e.g. bug fixes in fetchers)
3. are exceedingly difficult to implement for a small volunteer team without allowing for some breakage

I think this one used to be in (1), but I'm not convinced of that anymore.
(2) appears avoidable.

(3) does not really apply. The "cheap" solution of just rolling something out based on hopes and prayers would be completely irresponsible for very little benefit to the development cost. (Negative benefit if you include the risk of complicated workarounds for avoidable problems)

I'd suggest to treat this change as a project that's independent from flakes stabilisation, but I'm just a contributor and I'd like the @NixOS/nix-team to decide that instead. (I can join a meeting if you like)

**Comment by roberth** [comment]
Actually never mind!
This will be a **breaking fix** for the problem where relative path flakes duplicate the transitive inputs.
Current Nix versions will take the stale duplicates, whereas opting in to this change will cause the latest lock file from the relative path input to be used instead.

EDIT: but it's still a **fix** and flakes are experimental.

**Comment by eveeifyeve** [comment]
I am looking into this defined work further, but I am confused what I need to implement a poc.

Like you mention: `Design the follows detection algorithm`, Does this mean `inputs.<name>.inputs.<name>.follows = "";` or something else?


If it helps, I can understand visually, if you need to use a diagram feel free.

**Comment by adrian-gierakowski** [comment]
> Add `sparseNodes` field to root of lock file
> * referencing the existing nodes for their source pins

@roberth 

so you mean just a list of inputs from current flake.nix, but pinned?

**Comment by roberth** [comment]
> so you mean just a list of inputs from current flake.nix, but pinned?

**Yes, kind of**

Specifically a mapping from the declared inputs to the correct pins by means of references to `nodes`.
So if you have `inputs.a.inputs.b.url = "something"`,
then `sparseNodes` may have something like `{ "a/b": { "node": "b_2" } }`
and `nodes` has `"b_2": { ... }` (same value as before for backward compatibility).

**The set of locally pinned inputs is subtle**

"List" doesn't really do it justice. It's more of a tree, and which inputs are in it matters.

I've worked with the assumption that we want to "cache" inputs that are brought closer in using `follows`, and that this would be somewhat easy to achieve because we already have code that does this (and would remain involved during the migration/validation phases).
For example, you may have
```nix
inputs.a.follows = "b/nixpkgs";
inputs.b.url = ...;
```
Pinning only inputs with a `.url` would be sufficient, but it would slightly regress performance of evaluating `inputs.a` because that currently doesn't require fetching `b`.

Follows in the other direction usually don't incur a cost. For example:
```nix
inputs.a.url = ...;
inputs.b.inputs.nixpkgs.follows = "a";  # very common
```
`a` can be fetched with without fetching anything else first.

So what really matters is whether the rhs, e.g. "b/nixpkgs" is referred to from somewhere other than `b` or its descendants.
If it is, pinning locally yields a bit of an improvement.

**Comment by ruro** [comment]
> I've worked with the assumption that we want to "cache" inputs that are brought closer in using `follows`, and that this would be somewhat easy to achieve because we already have code that does this (and would remain involved during the migration/validation phases). For example, you may have
> 
> inputs.a.follows = "b/nixpkgs";
> inputs.b.url = ...;

As we all know, there are somewhere between 2 and N+1 hard problems in computer science, and at least one of them is cache invalidation.

In this case, caching the value of `a` to a "resolved" version of `b/nixpkgs` would not work as expected, if `b` is a relative path inside the same repo as `self`. So, in the case of

```nix
inputs.a.follows = "b/nixpkgs";
inputs.b.url = "path:./any-relative-path";
```

caching `b/nixpkgs` would be incorrect, because any changes in `b` should be automatically reflected in `self` (since they are locked together as parts of the same git repo).

**Comment by eveeifyeve** [comment]
I forgot to put a comment in here, as agreed to me @tomberek, @roberth & many other people in yesterday's meeting we will be starting to play around with in flake-compat before implementing it in cpp nix.

**Comment by roberth** [comment]
> caching `b/nixpkgs` would be incorrect

Very good point! I did not include that aspect in my latest explanation.

Relative path flake inputs as currently implemented already suffer from going out of sync, and relative path input lock file reuse is actually a necessary step towards or perhaps the majority part of solving that problem.

Fetching a relative path flake is virtually instantaneous, so we don't have to cache (duplicate) this `b/nixpkgs`.
Similarly, a `follows = "b/c/nixpkgs"` can also be taken from `b`'s lock just fine, discharging the `b/` and following the same rules in the context of `b`.

**Comment by edolstra** [comment]
I've started work on sparse lock files based on this design: https://gist.github.com/edolstra/e702cfccb5169ae592daabd48bcc9209

**Comment by roberth** [comment]
> ```c++
> struct LockFileV8
> {
>     struct Lock
> ```

The other day I [defined a *pin*](https://github.com/NixOS/nix/pull/16156/changes#diff-0b24fd6192e18b5383243c1f8103af8d991e940598d7e50c53791c077fce5755) to be that which is called `Lock` here.
I.e. lock as a group of pins that's ideally exhaustive. Adopting pin as the term for the small locked units could clarify your spec and the code.
I think it also helps that you don't have to call the large thing a "file" even when in memory.

> std::unique_ptr<LockFileV8> locks; // only present if the input has no lock file

*a* lock file, I presume.

> We'll have a separate call-flake-v7.nix 

I'd structure the file as a library with functions for each, as well as shared functions where it makes sense.
This will also help the flake-compat implementation.
I'd start by factoring the existing `call-flake.nix`, extracting functions until it's library shaped. High value and very low risk if you take advantage of small, inherently correct steps.

> ### Fingerprinting
> We need to verify that if flake.lock cannot be written, the evaluation cache is disabled (or ensure that the contents of the unwritten flake.lock is hashed into the fingerprint).

Nowadays couldn't we just overlay the lock file onto the root source and give that combination a fingerprint that adds in the lock file hash?

> ### Lock file diffs no longer show transitive input changes

The lock file was never exhaustive in the first place. Commands for fetching the closure and/or diffing will restore it to the level where it was.
At the ecosystem level it will make things better because currently some users feel incentivized to avoid normal flake inputs so that their lock files stay small. I have transitively [given in](https://flake.parts/options/flake-parts-partitions.html#opt-partitions._name_.extraInputsFlake) to that pressure too.

> ### Reduced laziness
> In some cases, the new lock file format may require fetching more inputs than the old format. With the old format, inputs.nixpkgs.follows = "foo/nixpkgs" would not require fetching foo if nixpkgs is not used, but with the new format it does. However, this is rare in practice, since you typically only reuse transitive inputs of inputs that you actually use.

I assume that's fixable by eagerly fetching those kinds of follows and storing them similar to a `.url` override.

**Comment by edolstra** [comment]
Update: https://github.com/DeterminateSystems/nix-src/pull/579 refactors Nix to support the new lock file format, without adding it yet:
* `LockedFlake` is now an abstract class (currently with only one concrete subclass, `LockedFlakeV7`). 
* All version-specific code is now contained in `src/libflake/lockfile-v7.cc`. This includes the V7 locking algorithm and the V7 graph representation of lock files (`Node`, `LockedNode`). The `LockedFlake` abstract interface exposes a tree rather than graph view (graphs are headache-inducing when it comes to updates).
* `call-flake.nix` no longer gets a JSON representation of the lock file; instead, it gets an abstract value (using `ExternalValue`) that wraps `LockedFlake`, and it has primops to (lazily) get the actual inputs by input attr path (e.g. `fetchFlakeInput lockedFlake ["a" "b" "c"]` gets the a/b/c input). So it won't need any changes to support the new format.

**Comment by edolstra** [comment]
> Nowadays couldn't we just overlay the lock file onto the root source and give that combination a fingerprint that adds in the lock file hash?

That's an interesting thought, since it would make `--override-input` work for people who do `builtins.fromJSON (builtins.readFile ./flake.lock)` (which [they probably really shouldn't do](https://github.com/NixOS/nix/issues/4639), but since it's currently allowed, it would be nice if it works correctly).

**Comment by roberth** [comment]
Having the locked flake be an ExternalValue seems like a bit of a smell to me.
The Nix language is perfectly capable of expressing all sorts of override machinery, while doing so in a purely functional way.
By doing it in C++, you
- don't provide a **purely functional semantics** of the dependency injection mechanism
- create a class of potential nondeterminism bugs that could have been ruled out systemically

The C++ should only output the lock nodes with perhaps some preapplied sourceInfos, but let the various overriding be done by `call-flake.nix`.
